File: Kawai codes/15.Random-Numbers/ch15pr03.m
%**************************************************************************
%*     Example 15.3                                                       *
%*     filename: ch15pr03.m                                               *
%*     program listing number: 15.3                                       *
%*                                                                        *
%*     This program generates Gaussian distributed random numbers         *
%*     using the Box-Muller method.                                       *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course.       *
%*     Last modification:  02/25/2017.                                    *
%**************************************************************************
clear all

N=10000000;
x = rand(N,2);
y(:,1) = sqrt(-2*log(x(:,1))).*cos(2*pi*x(:,2));
y(:,2) = sqrt(-2*log(x(:,1))).*sin(2*pi*x(:,2));
ymin=min(y(:));
ymax=max(y(:));
fprintf('the largest deviation=%f, %f\n',ymin,ymax)

h=histogram(y,201,'Normalization','pdf');
hold on
% true normal distribution
K=201;
xmin=floor(ymin);
xmax=ceil(ymax);
X=linspace(xmin,xmax,K);
F=exp(-X.^2/2)/sqrt(2*pi);
p=plot(X,F);
set(p,'color','red','linewidth',2)
xlabel('x','fontsize',14)
ylabel('\rho(x)','fontsize',14)
legend('Box-Muller','Exact')
axis([xmin xmax 0.0 0.5])
hold off

File: Kawai codes/15.Random-Numbers/ch15pr03.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
%**************************************************************************
%*     Example 16.3                                                       *
%*     filename: ch16pr03.m                                               *
%*     program listing number: 16.3                                       *
%*                                                                        *
%*     This program generates Gaussian distributed random numbers         *
%*     using the Box-Muller method.                                       *
%*                                                                        *
%*     Programed by Ryoichi Kawai for Computational Physics Course.       *
%*     Last modification:  02/15/2015.                                    *
%**************************************************************************
"""
import numpy as np
import matplotlib.pyplot as plt

N=10000000
x=np.random.rand(N,2)
y=np.zeros((N,2))
y[:,0] = np.sqrt(-2.0*np.log(x[:,0]))*np.cos(2.0*np.pi*x[:,1])
y[:,1] = np.sqrt(-2.0*np.log(x[:,0]))*np.sin(2.0*np.pi*x[:,1])
y=y.reshape(2*N)
ymin=min(y)
ymax=max(y)
print('the largest deviation={0:f}, {1:f}\n'.format(ymin,ymax))

K=201
plt.figure(figsize=(6,5))
plt.hist(y,K,normed=1,label='Box-Muller')
xmin=np.floor(ymin)
xmax=np.ceil(ymax)
X=np.linspace(xmin,xmax,K)
F=np.exp(-X**2/2.)/np.sqrt(2.*np.pi)
plt.plot(X,F,'-r',label='Exact')
plt.xlabel('x',fontsize=14)
plt.ylabel('p(x)',fontsize=14)
plt.legend(loc=1)
plt.show()
